### PR TITLE
Support AWS v4 provider

### PR DESCRIPTION
**Pull request by Nuru**
## what
- Migrate to AWS v4 Terraform provider
- Add features
  - Allow full S3 storage lifecycle configuration
  - Allow multiple bucket policy documents
  - Allow specifying the bucket name directly, rather than requiring it to be generated by `null-label`
  - Allow specifying S3 object ownership
  - Allow enabling S3 bucket keys for encryption
- Deprecate variable by variable specification of a single storage lifecycle rule
- Add extra safety measure `force_destroy_enabled`

## why
- AWS v4 broke this module
- Feature parity
- Replaced with more power and more flexible input
- Reduce the chance that automated upgrades will cause data loss

## references
- Upgrade instructions: [v0.27.0 to v0.28.0](https://github.com/cloudposse/terraform-aws-s3-log-storage/wiki/Upgrading-to-v0.28.0-and-AWS-provider-v4-(POTENTIAL-DATA-LOSS))
- Upgrade instructions: [versions prior to v0.27.0 to v0.27.0
](https://github.com/cloudposse/terraform-aws-s3-log-storage/wiki/Upgrading-to-v0.27.0-(POTENTIAL-DATA-LOSS))

### CODE REVIEW

File: examples/complete/variables.tf [hunk 1]
```diff
@@ -6,4 +6,10 @@ variable "allow_ssl_requests_only" {
   type        = bool
   default     = true
   description = "Set to `true` to require requests to use Secure Socket Layer (HTTPS/SSL). This will explicitly deny access to HTTP requests"
-}
\ No newline at end of file
+}
+
+variable "lifecycle_configuration_rules" {
+  type        = any
```
Comment (aknysh) [inline examples/complete/variables.tf:12]: should it be `list(any)` ?

File: main.tf [hunk 2]
```diff
@@ -45,26 +22,59 @@ moved {
...
   acl                = var.acl
-  force_destroy      = var.force_destroy
-  policy             = var.policy
+  force_destroy      = local.force_destroy
   versioning_enabled = var.versioning_enabled
 
-  lifecycle_rule_ids = [module.this.id]
-  lifecycle_rules    = [local.lifecycle_rule]
+  source_policy_documents = var.source_policy_documents
+  # Support deprecated `policy` input
```
Comment (aknysh) [inline main.tf:62]: this deprecated input actually causes permanent drift with `terraform plan`.
On internet, there are many mentions that the input must not be used at all.

File: variables-deprecated.tf [hunk 3]
```diff
@@ -0,0 +1,102 @@
+variable "policy" {
+  type        = string
+  description = "(Deprecated, use `source_policy_documents` instead): A valid bucket policy JSON document."
+  default     = ""
+}
+
+
```
Comment (aknysh) [inline variables-deprecated.tf:7]: remove the extra empty line

**Review by aknysh — COMMENTED** [review]
please see comments

File: main.tf (hunk 2 above)
Comment (Nuru) [reply main.tf:62]: @aknysh These are inputs to our `s3-bucket` module, not the AWS resource. Not a cause of drift.

File: examples/complete/variables.tf (hunk 1 above)
Comment (Nuru) [reply examples/complete/variables.tf:12]: Really it should match the module's variable definition, but I didn't want to have to keep them in sync.

File: main.tf (hunk 2 above)
Comment (aknysh) [reply main.tf:62]: not saying we need to remove this in this PR, but we need to review the bucket module

**Review by aknysh — APPROVED** [review]

**Pull request merged**